src/celeritas/optical/surface/SurfaceTraversalView.hh

```diff
@@ -0,0 +1,235 @@
+//------------------------------- -*- C++ -*- -------------------------------//
+// Copyright Celeritas contributors: see top-level COPYRIGHT file for details
+// SPDX-License-Identifier: (Apache-2.0 OR MIT)
+//---------------------------------------------------------------------------//
+//! \file celeritas/optical/surface/SurfaceTraversalView.hh
+//---------------------------------------------------------------------------//
+#pragma once
+
+#include "corecel/math/ArrayUtils.hh"
+#include "celeritas/optical/Types.hh"
+
+#include "SurfacePhysicsData.hh"
+#include "SurfacePhysicsUtils.hh"
+
+namespace celeritas
+{
+namespace optical
+{
+//---------------------------------------------------------------------------//
+/*!
+ * Manage one-dimensional logic for traversing an optical surface.
+ */
+class SurfaceTraversalView
+{
+  public:
+    //!@{
+    //! \name Type aliases
+    using SurfaceParamsRef = NativeCRef<SurfacePhysicsParamsData>;
+    using SurfaceStateRef = NativeRef<SurfacePhysicsStateData>;
+    //!@}
+
+    struct Initializer
+    {
+    };
+
+  public:
+    // Create view from surface physics data and state
+    inline CELER_FUNCTION SurfaceTraversalView(SurfaceParamsRef const&,
+                                               SurfaceStateRef const&,
+                                               TrackSlotId);
+
+    // Initialize track state
+    inline CELER_FUNCTION SurfaceTraversalView& operator=(Initializer const&);
+
+    // Whether the track is in the pre-volume
+    inline CELER_FUNCTION bool in_pre_volume() const;
+
+    // Whether the track is in the post-volume
+    inline CELER_FUNCTION bool in_post_volume() const;
+
+    // Whether track is currently moving off the surface
+    inline CELER_FUNCTION bool is_exiting() const;
+
+    // Whether track is in pre-/post-volume and direction is off the surface
+    inline CELER_FUNCTION bool is_exiting(SubsurfaceDirection) const;
+
+    // Position of the track in the surface crossing
+    inline CELER_FUNCTION SurfaceTrackPosition position() const;
+
+    // Next position of the track in the given direction
+    inline CELER_FUNCTION SurfaceTrackPosition next_position() const;
+
+    // Set position of the track in the surface crossing
+    inline CELER_FUNCTION void position(SurfaceTrackPosition);
+
+    // Number of valid positions of the track in the surface crossing
+    inline CELER_FUNCTION SurfaceTrackPosition::size_type num_positions() const;
+
+    // Current track traversal direction
+    inline CELER_FUNCTION SubsurfaceDirection direction() const;
+
+    // Set track traversal direction
+    inline CELER_FUNCTION void direction(SubsurfaceDirection);
+
+    // Cross subsurface interface in the given direction
+    inline CELER_FUNCTION void cross_interface(SubsurfaceDirection);
```

Comment:
Should we say `cross_subsurface` and rename the existing geometry method `cross_boundary` to `cross_surface`? (I think this would improve code clarity and we can do it later.)